src/fsharp/NameResolution.fs

```diff
@@ -2483,11 +2499,81 @@ let ResolveFieldPrim (ncenv:NameResolver) nenv ad typ (mp,id:Ident) =
         if isAppTy g typ then 
             match ncenv.InfoReader.TryFindRecdOrClassFieldInfoOfType(id.idText,m,typ) with
             | Some (RecdFieldInfo(_,rfref)) -> [ResolutionInfo.Empty, FieldResolution(rfref,false)]
-            | None -> error(Error(FSComp.SR.nrTypeDoesNotContainSuchField((NicePrint.minimalStringOfType nenv.eDisplayEnv typ), id.idText),m))
+            | None ->
+                // record label doesn't belong to record type
+                // predict other labels of same record
+                let typeName = NicePrint.minimalStringOfType nenv.eDisplayEnv typ
+                match typeName with
+                | "unit" | "int []" -> // TODO: match all non-record types - but how!?
+                    error(Error(FSComp.SR.nrTypeDoesNotContainSuchField(typeName, id.idText),m))
+                | _ ->                
+                    let labelsOfPossibleRecord =
+                        nenv.eFieldLabels
+                        |> Seq.filter (fun kv -> 
+                            kv.Value 
+                            |> List.map (fun r -> r.TyconRef.ToString()) // TODO: Ask Don how to compare type names here
+                            |> List.exists ((=) typeName))
+                        |> Seq.map (fun kv -> kv.Key)
+                        |> Set.ofSeq
+
+                    let givenFields = allFields |> List.map (fun fld -> fld.idText) |> List.filter ((<>) id.idText) |> Set.ofList
+
+                    let predictedLabels = Set.difference labelsOfPossibleRecord givenFields
+
+                    let details =
+                        ErrorResolutionHints.CreatePredictionText
+                            FSComp.SR.undefinedNameRecordLabelDetails
+                            id.idText
+                            predictedLabels
+
+                    let (l,errorText) = FSComp.SR.nrRecordDoesNotContainSuchLabel(typeName, id.idText)
+                    let errorText =
+                        match details with
+                        | None -> l,errorText
+                        | Some details -> l,errorText + System.Environment.NewLine + details
+                    error(Error(errorText,m))
         else 
             let frefs = 
                 try Map.find id.idText nenv.eFieldLabels 
-                with :? KeyNotFoundException -> error (UndefinedName(0,FSComp.SR.undefinedNameRecordLabel,id,NameMap.domainL nenv.eFieldLabels))
+                with :? KeyNotFoundException ->
```

Comment:
Q: would it be wise to change that thing to Map.tryFind instead catching the exn?